server/src/main/java/keywhiz/service/daos/SecretDAO.java

```diff
@@ -198,6 +198,33 @@ public long createOrUpdateSecret(String name, String encryptedSecret, String cre
     return ImmutableList.copyOf(results);
   }
 
+  /**
+   * @param name of secret series to look up secrets by.
+   * @param newestIdx the first index to select in a list of versions sorted by creation time
+   * @param oldestIdx the last index to select in a list of versions sorted by creation time
+   * @return Versions of a secret matching input parameters or Optional.absent().
+   */
+  public Optional<SecretSeriesAndVersions> getSecretVersionsByName(String name, int newestIdx,
+      int oldestIdx) {
+    checkArgument(!name.isEmpty());
+    checkArgument(newestIdx >= 0);
+    checkArgument(oldestIdx >= 0);
+
```

Comment:
What is the behavior if I ask for newest=10 oldest=1? You should probably have a check requiring oldest > newest, if oldest != 0. Otherwise, looking at the query in getSecretVersionsBySecretId, I believe you end up with a negative value in the limit clause, which will either result in an exception from jooq or an error from mysql.